components/infra-proxy-service/migrations/pipeline/utility.go

```diff
@@ -305,39 +307,185 @@ func Unzip(ctx context.Context, mst storage.MigrationStorage, result pipeline.Re
 		}
 	}
 
 	result.Meta.UnzipFolder = filepath.Dir(fpath)
-	reader.Close()
+
+	_ = reader.Close()
+
+	_ = reader.Close()
 	_, err = mst.CompleteUnzip(ctx, result.Meta.MigrationID, result.Meta.ServerID, 0, 0, 0)
 	if err != nil {
 		log.Errorf("Failed to update status in DB: %s :%s", result.Meta.MigrationID, err)
 	}
 	return result, nil
 }
+
+func GetUsersForBackup(ctx context.Context, st storage.Storage, mst storage.MigrationStorage, result pipeline.Result) (pipeline.Result, error) {
+	log.Info("starting with user parseing phase for migration id: ", result.Meta.MigrationID)
+
+	file := path.Join(result.Meta.UnzipFolder, "key_dump.json")
+
+	keyDumpByte, err := ioutil.ReadFile(file)
+	if err != nil {
+		log.Errorf("failed to read keydump file for user parsing for the migration id: %s : %s", result.Meta.MigrationID, err.Error())
+		_, _ = mst.FailedUsersParsing(ctx, result.Meta.MigrationID, result.Meta.ServerID, err.Error(), 0, 0, 0)
+		return result, err
+	}
+
+	var keyDumps []pipeline.KeyDump
+	if err := json.Unmarshal(keyDumpByte, &keyDumps); err != nil {
+		log.Errorf("failed to unmarshal for user parsing for the migration id: %s : %s", result.Meta.MigrationID, err.Error())
+		_, _ = mst.FailedUsersParsing(ctx, result.Meta.MigrationID, result.Meta.ServerID, err.Error(), 0, 0, 0)
+		return result, err
+	}
+
+	serverUsers := keyDumpTOUser(keyDumps)
+	automateUsers, err := st.GetUsers(ctx, result.Meta.ServerID)
+	if err != nil {
+		log.Errorf("failed to get users form db for user parsing for the migration id: %s : %s", result.Meta.MigrationID, err.Error())
+		_, _ = mst.FailedUsersParsing(ctx, result.Meta.MigrationID, result.Meta.ServerID, err.Error(), 0, 0, 0)
+		return result, err
+	}
+	var mappedUsers []pipeline.User
+	checked := map[string]bool{}
+
+	mappedUsers = append(mappedUsers, SkipUpdateUser(serverUsers, automateUsers, checked)...)
+	mappedUsers = append(mappedUsers, DeleteUser(serverUsers, automateUsers, checked)...)
+	mappedUsers = append(mappedUsers, InsertUser(serverUsers, checked)...)
+
+	result.ParsedResult.Users = mappedUsers
+	return result, nil
+}
+
+// Clean serialized_object and Polulate Users struct
+func keyDumpTOUser(keyDump []pipeline.KeyDump) []pipeline.User {
+	var users []pipeline.User
+	for _, kd := range keyDump {
+		sec := map[string]string{}
+		if err := json.Unmarshal([]byte(kd.SerializedObject), &sec); err != nil {
+			log.Errorf("failed to pasre user's first, middle and last name: %s", err.Error())
+		}
+		user := pipeline.User{
+			Username:    kd.Username,
+			Email:       kd.Email,
+			DisplayName: sec["display_name"],
+			FirstName:   sec["first_name"],
+			LastName:    sec["last_name"],
+			MiddleName:  sec["middle_name"],
+		}
+		users = append(users, user)
+	}
+	return users
+}
+
+// SkipUpdateUser
+func SkipUpdateUser(serverUser []pipeline.User, automateUser []storage.User, checked map[string]bool) []pipeline.User {
+	var parsedUsers []pipeline.User
+
+	for _, aUser := range automateUser {
```

Comment:
This is a very inefficient way of finding and handling comparing two set of users. Should change it to a map so that a single loop can work